src/modules/span_modules.py

```diff
@@ -0,0 +1,210 @@
+# Implementation of span classification modules
+
+import torch
+import torch.nn as nn
+import numpy as np
+import torch.nn.functional as F
+from torch.autograd import Variable
+
+from ..tasks.tasks import Task
+from .import modules
+
+from allennlp.modules.span_extractors import \
+    EndpointSpanExtractor, SelfAttentiveSpanExtractor
+from allennlp.nn.util import move_to_device, device_mapping
+
+from typing import Dict, Iterable, List
+
+import logging as log
+
+class SpanClassifierModule(nn.Module):
+    ''' 
+        Build span classifier components as a sub-module.
+        Classifier that allows for spans and text as input.
+        Use same classifier code as build_single_sentence_module,
+        except we'll use span indices to extract span representations, 
+        and use these as input to the classifier.
+    '''
+
+    def _make_span_extractor(self):
+        if self.span_pooling == "attn":
+            return SelfAttentiveSpanExtractor(self.proj_dim)
+        else:
+            return EndpointSpanExtractor(self.proj_dim,
+                                         combination=self.span_pooling)
+
+    def _make_cnn_layer(self, d_inp):
+        """
+        Make a CNN layer as a projection of local context.
+        CNN maps [batch_size, max_len, d_inp]
+        to [batch_size, max_len, proj_dim] with no change in length.
+        """
+        k = 1 + 2 * self.cnn_context
+        padding = self.cnn_context
+        return nn.Conv1d(d_inp, self.proj_dim, kernel_size=k,
+                         stride=1, padding=padding, dilation=1,
+                         groups=1, bias=True)
+
+    def __init__(self, task, d_inp: int, task_params, num_spans=2):
+        assert num_spans > 0, "Please set num_spans to be more than 0"
+        super(SpanClassifierModule, self).__init__()
+        # Set config options needed for forward pass.
+        self.loss_type = task_params['cls_loss_fn']
+        self.span_pooling = task_params['cls_span_pooling']
+        self.single_sided = task.single_sided
+        self.cnn_context = task_params.get('cnn_context', 0)
+        self.num_spans = num_spans
+        self.proj_dim = task_params['d_hid']
+        self.projs = []
+
+        for i in range(num_spans):
+            # this effectively acts as a linear layer (if cnn_context = 0)
+            proj = self._make_cnn_layer(d_inp).cuda() \
+                if torch.cuda.is_available() else self._make_cnn_layer(d_inp)
+            self.projs.append(proj)
+        self.span_extractors = []
+
+        # This can be Lee's self-pooling operator (https://arxiv.org/abs/1812.10860)
+        # or any of the other types of span pooling operators
```

Comment:
Not sure what 'this' refers to. If it's for span_extractor, maybe put the comment in _make_span_extractor?